src/components/Table/utils/TableState.ts

```diff
@@ -218,25 +218,61 @@ export class TableState {
   setColumns(columns: GridColumnWithId<any>[], visibleColumnsStorageKey: string | undefined): void {
     const isInitial = !this.columns || this.columns.length === 0;
     if (columns !== this.columns) {
-      this.columns = columns;
       this.visibleColumnsStorageKey = visibleColumnsStorageKey ?? camelCase(columns.map((c) => c.id).join());
       this.visibleColumns.replace(readOrSetLocalVisibleColumnState(columns, this.visibleColumnsStorageKey));
-      const expandedColumnIds = columns.filter((c) => c.initExpanded).map((c) => c.id);
-      if (isInitial) this.expandedColumns.replace(expandedColumnIds);
+      // list of local storage columns
+      const localStorageColumns = this.persistCollapse ? readExpandedColumnsStorage(this.persistCollapse) : [];
+      // expandedColumnIds initialized as localStorageColumns
+      const existingExpandedColumnIds = isInitial ? localStorageColumns : this.expandedColumnIds;
+
+      // list of all existing columns
+      const existingColumnIds = this.columns.map((c) => c.id);
+
+      // filter to find new columns that should be initially expanded
+      const newExpandedColumnsIds = columns
+        .filter((c) => !existingColumnIds.includes(c.id) && c.initExpanded)
+        .map((c) => c.id);
+
+      const isDifferent = !newExpandedColumnsIds.every((c) => existingExpandedColumnIds.includes(c));
+      if (isDifferent) {
+        existingExpandedColumnIds.push(...newExpandedColumnsIds);
+      }
+
+      const newExpanded = columns.filter((c) => existingExpandedColumnIds.includes(c.id));
+      // if we have existing expanded columns ids, then do a Promise.all() and execute column.expandColumns
+      Promise.all(
+        newExpanded.map(async (c) => {
+          // if the column is initExpanded, manually toggle the column
+          if (c.initExpanded) {
+            this.toggleExpandedColumn(c.id);
+          }
```

Comment:
**thought:** A couple thoughts here:
1. Instead of calling the `toggle`, I think we should be more explicit by doing a `this.expandedColumns.add(c.id)`
2. Though, I also don't think we want to make this call multiple times. Instead we should do a single `replace`... which I think is already happening in the `then` clause... 🤔